prow/kube/prowjob_collector.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kube
+
+import (
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+type prowJobClient interface {
+	List(opts metav1.ListOptions) (*prowapi.ProwJobList, error)
+}
+
+//https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector
+type prowJobCollector struct {
+	name     string
+	client   prowJobClient
+	selector string
+}
+
+func (pjc prowJobCollector) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
Can we implement a different descriptor? Seems weird that we would leave this empty

Reply:
Basically we need to send nothing to `ch`.
As far as I understood, any descriptor sent to `ch` will be used to check the data later.
Not sure if I could do something meaningful here.

Reply:
Ah, ok, that makes sense. Could we add that to the comment? That part was not clear to me

Reply:
Maybe something like:


```
// Normally, we'd send descriptions into this channel. However, we cannot do so for these 
// metrics as their label sets are dynamic. This is a ake-our-own-risk action and also a
// compromise for implementing a metric with both dynamic keys and dynamic values in
// the label set. 
// https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
```

Reply:
Done.